include/exadg/convection_diffusion/driver.cpp

```diff
@@ -188,28 +187,152 @@ Driver<dim, Number>::ale_update() const
+                                   pde_operator->get_constraints(),
+                                   *pde_operator->get_mapping(),
+                                   solution,
+                                   application->get_parameters().degree +
+                                     1 /* n_face_quadrature_points */,
+                                   application->get_parameters().amr_data);
+}
+
+template<int dim, typename Number>
+void
+Driver<dim, Number>::setup_after_coarsening_and_refinement()
+{
+  pde_operator->initialize_dof_handler_and_constraints();
+  pde_operator->setup();
+
+  postprocessor->setup(*pde_operator);
+
+  VectorType const * velocity_ptr = nullptr;
+  VectorType         velocity;
+
+  if(application->get_parameters().problem_type == ProblemType::Unsteady)
+  {
+    if(application->get_parameters().temporal_discretization == TemporalDiscretization::BDF)
+    {
+      std::shared_ptr<TimeIntBDF<dim, Number>> time_integrator_bdf =
+        std::dynamic_pointer_cast<TimeIntBDF<dim, Number>>(time_integrator);
+
+      if(application->get_parameters().get_type_velocity_field() == TypeVelocityField::DoFVector)
+      {
+        Assert(false, dealii::ExcMessage("ALE and AMR combination not implemented."));
+        pde_operator->initialize_dof_vector_velocity(velocity);
+        pde_operator->interpolate_velocity(velocity, time_integrator->get_time());
+        velocity_ptr = &velocity;
+      }
+
+      auto & grid = *application->get_grid_non_const();
+      grid.coarse_triangulations =
+        dealii::MGTransferGlobalCoarseningTools::create_geometric_coarsening_sequence(
+          *application->get_grid_non_const()->triangulation);
+
+      pde_operator->setup_solver(time_integrator_bdf->get_scaling_factor_time_derivative_term(),
+                                 velocity_ptr);
+    }
+    else
+    {
+      AssertThrow(application->get_parameters().temporal_discretization ==
+                    TemporalDiscretization::ExplRK,
+                  dealii::ExcMessage("Not implemented."));
+    }
+  }
+  else
+  {
+    AssertThrow(false, dealii::ExcMessage("Not implemented"));
+  }
+}
+
+
+template<int dim, typename Number>
+void
+Driver<dim, Number>::do_adaptive_refinement(unsigned int const time_step_number)
+{
+  dealii::Triangulation<dim> & tria = *application->get_grid_non_const()->triangulation;
+
+  // AMR is only implemented for implicit timestepping.
+  if(application->get_parameters().temporal_discretization != TemporalDiscretization::BDF)
+  {
+    AssertThrow(false, dealii::ExcNotImplemented());
+  }
+  else
+  {
+    std::shared_ptr<TimeIntBDF<dim, Number>> bdf_time_integrator =
+      std::dynamic_pointer_cast<TimeIntBDF<dim, Number>>(time_integrator);
+
+    if(trigger_coarsening_and_refinement_now(application->get_parameters().amr_data,
+                                             time_step_number))
+    {
+      mark_cells_coarsening_and_refinement(tria, bdf_time_integrator->get_solution_np());
```

Comment:
In my opinion, just before this function call we should assert that the above dynamic cast was successful. Then, the above if-else-branch for BDF time integration with corresponding assert can be removed and the check appears more locally, i.e. only for the functionality where time_integrator vs. bdf_time_integrator is relevant.